### ISSUE TITLE
Injected `const` inlined as `var`

### ISSUE THREAD

**Issue by steveluscher** [evanw/esbuild#2695]
### `./shim.ts`

```ts
export const __DEV__ = process['en' + 'v'].NODE_ENV === 'development';
```

### `./index.ts`

```ts
if (__DEV__) {
  const foo = 'bar';
  console.debug(foo);
}
```

### `dist/index.js`

```
// shim.ts
var __DEV__ = process["env"].NODE_ENV === "development"

// index.ts
if (__DEV__) {
  const foo = 'bar';
  console.debug(foo);
}
```

I would like the shimmed `__DEV__` to be `const` instead of `var`.

**Comment by evanw** [comment]
> I would like the shimmed `__DEV__` to be `const` instead of `var`.

Why is that?

**Comment by steveluscher** [comment]
Some dead code eliminators look at the mutable binding and refuse to reap the dead branches when `__DEV__` is `false`.

**Comment by hyrious** [comment]
I suppose this issue is about making a library, where esbuild's build will make some DCE to not happen.

1. Keep `process.env.NODE_ENV` in the output. This is the ideal way to make the esm bundle shrink in production mode. An example is vue's esm export. https://cdn.jsdelivr.net/npm/vue@3.2.45/dist/vue.esm-bundler.js

2. esbuild has an optimization that top-level constants can be inlined when they're not exported. But it also converts `const` to `var` in order to prevent TDZ issue. This is ok when you're bundling the final work, but not very ok in library. For example,

   ```js
   // in some lib
   export const FOO = 1
   // becomes var FOO = 1

   // in user peoject
   import { FOO } from 'lib'
   // no inline const happen any more
   ```

**Comment by steveluscher** [comment]
It is about making a library, @hyrious; I should have mentioned that.

Before I found `inject`, I tried this:

```ts
define: {
  __DEV__: 'process.env.NODE_ENV === "development"',
}
```

That didn't work because define's values have to be JSON.

```shell
error: Invalid define value (must be an entity name or valid JSON syntax): process.env.NODE_ENV === "development"
```

I would otherwise be happy to keep `process.env.NODE_ENV` in the output, but I was hoping to let my library's contributors simply use a compile-time macro (`__DEV__`) rather than to type out `process.env.NODE_ENV === "development"` or `process.env.NODE_ENV !== "production"`.

Vue does this too. They achieve the replacement that I hoped to achieve [using Rollup's ‘replace’ plugin](https://github.com/vuejs/vue/blob/9dd006b481b4299462e044741bac0861c0b1775c/scripts/config.js#L272-L290).

If esbuild would just honor the const/let/var binding that I exported in `./shim.ts`, then I think I'll be good. I'd also take `define` working more like `replace`, or an actual `replace` option.

**Comment by rizrmd** [comment]
Actually, you can manually replace `var __DEV__ =` with `const __DEV__ =` after esbuild done building. It's not ideal, but that's what I do.

**Comment by steveluscher** [comment]
Indeed. I was hoping to avoid adding a build step on top of the… build step.

**Comment by maxpatiiuk** [comment]
> Actually, you can manually replace var __DEV__ = with const __DEV__ = (using plain old .replace('var __DEV__ =', 'const __DEV__ =')) after esbuild done building. It's not ideal, but that's what I do.

Not ideal - the source map is out of order if you do that, and it can wrongly replace references in strings/comments too
